File: python/rx-test.py
from TAP.Simple import *
import Rx
import json
import re
import pdb
import os

plan(None)
os.chdir('..')
rx = Rx.Factory({ "register_core_types": True });

isa_ok(rx, Rx.Factory)

index = json.loads(open('spec/index.json').read())

test_data     = {}
test_schemata = {}

def normalize(entries, test_data):
  if entries == '*':
    entries = { "*": None }

  if type(entries) is type([]):
    new_entries = { }
    for n in entries: new_entries[n] = None
    entries = new_entries

  if len(entries) == 1 and '*' in entries:
    value = entries["*"]
    entries = { }
    for k in test_data.keys(): entries[k] = value

  return entries

for filename in index:
  if filename == 'spec/index.json': continue
  payload = json.loads(open(filename).read())

  parts = filename.split('/')
  parts.pop(0)

  leaf_name = '/'.join(parts[1:])
  leaf_name = re.sub('\.json$', '', leaf_name)

  filetype = parts.pop(0)

  if filetype == 'schemata':
    test_schemata[ leaf_name ] = payload
  elif filetype == 'data':
    test_data[ leaf_name ] = {}

    if type(payload) is type([]):
      for data_str in payload:
        boxed_data = json.loads("[ %s ]" % data_str)
        test_data[ leaf_name ][ data_str ] = boxed_data[0]

    else:
      for entry in payload.keys():
        boxed_data = json.loads("[ %s ]" % payload[entry])
        test_data[ leaf_name ][ entry ] = boxed_data[0]
  else:
    raise StandardError("weird file in data dir: %s" % filename)

schema_names = list(sorted(test_schemata.keys()))

for schema_name in schema_names:
  rx = Rx.Factory({ "register_core_types": True });

  schema_test_spec = test_schemata[ schema_name ]

  if schema_test_spec.get("composedtype", False):
    try:
      rx.learn_type(schema_test_spec['composedtype']['uri'],
                    schema_test_spec['composedtype']['schema'])
    except Rx.SchemaError as e:
      if schema_test_spec['composedtype'].get("invalid", False):
        ok(1, "BAD COMPOSED TYPE: schemata %s" % schema_name)
        continue
      else:
        raise

    if schema_test_spec['composedtype'].get("invalid", False):
      ok(0, "BAD COMPOSED TYPE: schemata %s" % schema_name)

    if schema_test_spec['composedtype'].get("prefix", False):
       rx.add_prefix(schema_test_spec['composedtype']['prefix'][0],
                     schema_test_spec['composedtype']['prefix'][1])

  try:
    schema = rx.make_schema(schema_test_spec["schema"])
  except Rx.SchemaError as e:
    #pdb.set_trace()
    if schema_test_spec.get("invalid", False):
      ok(1, "BAD SCHEMA: schemata %s" % schema_name)
      continue
    else:
      raise

  if schema_test_spec.get("invalid", False):
    ok(0, "BAD SCHEMA: schemata %s" % schema_name)
    continue

  if not schema: raise StandardError("got no schema obj for valid input")

  for pf in [ 'pass', 'fail' ]:
    for source in schema_test_spec.get(pf, []):
      to_test = schema_test_spec[pf][ source ]

      to_test = normalize(to_test, test_data[ source ])
      # if to_test == '*': to_test = test_data[ source ].keys()

      for entry in to_test:
        result = None
        try:
          schema.validate(test_data[source][entry])
          result = True
        except Rx.SchemaMismatch as e:
          print(str(e))
          result = False

        desc = "%s/%s against %s" % (source, entry, schema_name)

        if pf == 'pass':
          ok(result, "VALID  : %s" % desc)
          #if not result:
          #  pdb.set_trace()
          #  result = schema.check( test_data[source][entry] )
        else:
          ok(not result, "INVALID: %s" % desc)
          #if result:
          #  pdb.set_trace()
          #  result = schema.check( test_data[source][entry] )

File: php/RxTest.php
#!/usr/bin/php
<?php
include_once("vendor/autoload.php");

require 'ext/Test.php';

$index_json = file_get_contents('spec/index.json');
$index = json_decode($index_json);

$test_data = array();
$test_schemata = array();

foreach ($index as $file) {
  if ($file == 'spec/index.json') continue;
  $parts = explode('/', $file);
  
  array_shift($parts);
  $type = array_shift($parts);

  $leaf = $parts[ count($parts) - 1 ];
  $leaf = preg_replace('/\.json$/', '', $leaf);

  if ($type == 'schemata') {
    $file = join("/", $parts);
    $test_schemata[ $leaf ] = json_decode(
      file_get_contents("spec/schemata/$file")
    );
  } else if ($type == 'data') {
    $file = join("/", $parts);
    $test_data[ $leaf ] = json_decode(
      file_get_contents("spec/data/$file")
    );
  } else {
    die("unknown entries in index.json");
  }
}

foreach ($test_data as $k => $v) {
  if (is_array($v)) {
    $new_v = new stdClass();

    foreach ($v as $entry) {
      $new_v->$entry = $entry;
    }

    $test_data[$k] = $v = $new_v;
  }

  foreach ($test_data[$k] as $entry => $json) {
    $j_arr = json_decode("[ $json ]");
    $test_data[$k]->$entry = $j_arr[0];
    # diag("loaded $json ($k/$entry) as " . var_export($j_arr[0], true));
  }
}

function normalize($entries, $test_data) {
  if ($entries == '*') {
    $entries = new stdClass();
    $key = "*";
    $entries->$key = null;
  }

  if (is_array($entries)) {
    $new_entries = new stdClass();
    foreach ($entries as $entry) {
      $new_entries->$entry = $entry;
    }

    $entries = $new_entries;
  }

  if (count((array) $entries) == 1 and property_exists($entries, "*")) {
    $key  = "*";
    $value = $entries->$key;
    $new_entries = new stdClass();
    foreach ($test_data as $name => $entry) {
      $new_entries->$name = $entry;
    }
    $entries = $new_entries;
  }

  return $entries;
}

function test_json($x, $y) {
  return false;
}

asort($test_schemata);
foreach ($test_schemata as $schema_name => $test) {
  if (isset($_ENV["RX_TEST_SCHEMA"]) and $_ENV["RX_TEST_SCHEMA"] != $schema_name)
    continue;

  $Rx = new Rx\Rx();

  $schema = null;

  if (isset($test->composedtype)) {
    try {
      $Rx->learnType($test->composedtype->uri, $test->composedtype->schema);
    } catch (Exception $e) {
      if (isset($test->composedtype->invalid)) {
        pass("BAD COMPOSED TYPE: $schema_name");
        continue;
      } else {
        throw $e;
      }
    }

    if (isset($test->composedtype->invalid)) {
      fail("BAD COMPOSED TYPE: $schema_name");
      continue;
    }

    if (isset($test->composedtype->prefix))
      $Rx->addPrefix($test->composedtype->prefix[0],
                      $test->composedtype->prefix[1]);
  }

  try {
    $schema = $Rx->makeSchema($test->schema);
  } catch (Exception $e) {
    if (isset($test->invalid)) {
      pass("BAD SCHEMA: $schema_name");
      continue;
    } else {
      throw $e;
    }
  }

  if (isset($test->invalid)) {
    fail("BAD SCHEMA: $schema_name");
    continue;
  }

  if (! $schema) die("did not get schema for valid input");

  foreach (array('pass', 'fail') as $pf) {
    $expect = ($pf == 'pass') ? 'VALID  ' : 'INVALID';
    if (!isset($test->$pf)) continue;

    foreach ($test->$pf as $source => $entries) {
      $entries = normalize($entries, $test_data[$source]);

      foreach ($entries as $name => $want) {
        $value = $test_data[$source]->$name;

        $result = $schema->check($value);
        if ($pf == 'fail') $result = ! $result;

        if ("$source/$entry" == "num/0e0")
          todo_start("PHP's json_decode can't handle 0e0 as number");

        ok($result, "$expect: $source/$name against $schema_name");

        if ("$source/$entry" == "num/0e0")
          todo_end();
      }
    }
  }
}

?>
